### PR TITLE
Operator/CES: Simplify rate limit configuration

### PR DESCRIPTION
**Pull request by thorn3r**
The CiliumEndpointSlice controller has 6 rate limit related config options. This PR attempts to simplify configuration by reducing that set to 2: one for supplying a custom, dynamic rate limit config, and another to supply a base multiplier for simple tuning of an existing config.

Note: The supplied default dynamic rate limit may need some adjustment. This will likely be updated after future scale testing and/or PR feedback.

See individual commits for more details

```release-note
Simplify rate limit configuration options for the CiliumEndpointSlice controller.
```

### CODE REVIEW

**Review by tklauser — CHANGES_REQUESTED** [review]
I haven't looked at the entire PR in detail but one thing that stood out on initial review is that the old options and Helm values are removed rather than deprecated.

I think our current policy is to deprecate config options/values in Cilium release v1.n and and introduce the new options/values in addition. The deprecation of the options/values is documented in the v1.n release upgrade section.

In Cilium release v1.n+1 we'd then remove the deprecated options/values and solely rely on the new configuration and again document the removal.

I understand this process makes this PR a bit more complicated to implement but will allow users to upgrade gradually to the new options rather than on upgrade breaking their existing configuration relying on the old options/values.

**Comment by thorn3r** [comment]
@tklauser is that the case for beta features as well? The biggest impact here would probably be the helm change, since there are probably few users with custom rate limiting configured (the options are undocumented except for cmdref).

Agreed on deprecation vs removal for the helm change, and happy to do the same for the rate limit options as well if you think that's necessary

**Review by ghost — COMMENTED** [review]
Same as @tklauser I would expect the Helm value to be deprecated rather than removed, but I'll leave it to someone else to comment on practices for beta features.

Either way, could you document the Helm change in the upgrade notes (`Documentation/operations/upgrade.rst`)?

**Comment by thorn3r** [comment]
thanks @lambdanis! pushed the change to deprecate the helm value rather than remove it, and added it to the upgrade notes

**Comment by tklauser** [comment]
> @tklauser is that the case for beta features as well? The biggest impact here would probably be the helm change, since there are probably few users with custom rate limiting configured (the options are undocumented except for cmdref).

Ah, I see. I wasn't considering that this is only affecting a beta feature.

> Agreed on deprecation vs removal for the helm change, and happy to do the same for the rate limit options as well if you think that's necessary

If the options are for a beta feature and not documented outside of the cmdref I'd say it's probably okay to just deprecate the Helm values and replace the options. But I'm not entirely sure of the policy in that regard, so might be good to confirm with maintainers and/or in the #development channel on slack.

File: install/kubernetes/cilium/templates/validate.yaml
Comment (giorio94) [inline install/kubernetes/cilium/templates/validate.yaml]: Should account for both the new and the deprecated setting.

File: operator/pkg/ciliumendpointslice/endpointslice.go
Comment (giorio94) [inline operator/pkg/ciliumendpointslice/endpointslice.go]: These changes seem unrelated, and should be reverted.

File: operator/pkg/ciliumendpointslice/rate_limit.go
Comment (giorio94) [inline operator/pkg/ciliumendpointslice/rate_limit.go]: It's personally suggest to bail out loudly if the configuration cannot be parsed, to avoid surprises to users.

File: test/helpers/kubectl.go
Comment (giorio94) [inline test/helpers/kubectl.go]: Most of the changes in this file appear to be unrelated, and should be reverted.

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (giorio94) [inline install/kubernetes/cilium/values.yaml.tmpl]: I'd suggest to squash this commit with the first one, as strongly related.

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (giorio94) [inline install/kubernetes/cilium/values.yaml.tmpl]: Looking at the code, it seems that the maximum allowed values are respectively 50 and 100, which may be worth documenting.

File: operator/pkg/ciliumendpointslice/rate_limit.go
Comment (giorio94) [inline operator/pkg/ciliumendpointslice/rate_limit.go]: I personally find a complex default configuration like this to be possibly confusing for users, as there's no easy way to figure it out except looking at the code. How about having a single limit/burst default, and then overriding it by default through the helm values? That way, the default values would be immediately visible to the users, and tuning would also be way simpler. That approach would also allow to drop the additional multiplier flag introduced in the second commit, as the values could be tuned directly if necessary.

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (giorio94) [inline install/kubernetes/cilium/values.yaml.tmpl]: I'd personally strip the term `dynamic` from the various settings/flags, considering that it's the only rate limit configuration available.

File: Documentation/operations/upgrade.rst
Comment (giorio94) [inline Documentation/operations/upgrade.rst]: I'd be more explicit in the wording here, as the deprecated values seem to be effectively ignored with the current implementation, which is not what a user would likely expect reading *deprecated*. I personally don't have a strong opinion on preserving the existing functionality considering that CiliumEndpointSlices are a beta feature, but it should be stated clearly (I'd also ensure that the operator logs errors in case the deprecated flags are used if the values are ignored).

**Review by giorio94 — CHANGES_REQUESTED** [review]
Nice work, always in favor of simplifying the Cilium's configuration!

I've left a few suggestions inline, mainly about the possibility of making the default settings more explicit, further removing the multiplier setting, and clarifying the semantic around the *deprecated* wording.

File: operator/pkg/ciliumendpointslice/rate_limit.go
Comment (thorn3r) [reply operator/pkg/ciliumendpointslice/rate_limit.go]: SGTM :+1: I think that's a more user-friendly approach :+1:

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (thorn3r) [reply install/kubernetes/cilium/values.yaml.tmpl]: Added in the comments. I was hoping helm-docs supported docstrings for fields of an object within a list, but I think that may beyond what the tool can handle. Added details for the fields within the top-level comment and moved the default config to the helm values..

File: test/helpers/kubectl.go
Comment (thorn3r) [reply test/helpers/kubectl.go]: It seems I unknowingly had `gofumpt` enabled :sweat_smile:

File: operator/pkg/ciliumendpointslice/rate_limit.go
Comment (thorn3r) [reply operator/pkg/ciliumendpointslice/rate_limit.go]: Done :+1:

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (thorn3r) [reply install/kubernetes/cilium/values.yaml.tmpl]: Good point, updated to make this change

File: Documentation/operations/upgrade.rst
Comment (thorn3r) [reply Documentation/operations/upgrade.rst]: Done for both, let me know what you think

**Review by tklauser — APPROVED** [review]
Thanks @thorn3r!

**Review by ghost — APPROVED** [review]
docs good

File: operator/pkg/ciliumendpointslice/rate_limit.go
Comment (giorio94) [inline operator/pkg/ciliumendpointslice/rate_limit.go]: Nit. AFAIR it is better to return an error to hive, rather than calling Fatal directly, to enable proper shutdown.

File: operator/pkg/ciliumendpointslice/rate_limit.go
Comment (giorio94) [inline operator/pkg/ciliumendpointslice/rate_limit.go]: Nit. This check appears to be redundant.

File: operator/pkg/ciliumendpointslice/cell.go
Comment (giorio94) [inline operator/pkg/ciliumendpointslice/cell.go]: NIt. It may be worth to display the actual default configuration, rather than an empty string.

**Review by giorio94 — APPROVED** [review]
lgtm, thanks! Just a couple of final nits inline.

File: Documentation/operations/upgrade.rst
Comment (giorio94) [reply Documentation/operations/upgrade.rst]: :+1: Thanks!

**Review by joamaki — APPROVED** [review]

File: operator/pkg/ciliumendpointslice/rate_limit.go
Comment (thorn3r) [reply operator/pkg/ciliumendpointslice/rate_limit.go]: Updated the surrounding functions to bubble the error up to hive

File: operator/pkg/ciliumendpointslice/cell.go
Comment (thorn3r) [reply operator/pkg/ciliumendpointslice/cell.go]: I removed the default rate limit struct and put the default values here instead.

**Comment by thorn3r** [comment]
Changed the config parsing to use a decoder and disallow unknown fields to avoid unexpected behavior if a user enters a typo and the intended field remains 0. Added a short test for this as well.

**Review by aanm — APPROVED** [review]

**Pull request merged**
